c++/src/kj/common.h

```diff
@@ -1295,19 +1298,65 @@ class NullableValue {
   friend NullableValue<U>&& readMaybe(Maybe<U>&& maybe);
 };
 
+template <typename T, typename = EnableIf<!isLvalueReference<T>()>>
+class ExplicitPtr {
+  // This class wraps a raw pointer to T. It supports one-way implicit conversion: raw pointers can
+  // be implicitly converted to an ExplicitPtr, but an ExplicitPtr cannot be implicitly converted
+  // back to a raw pointer. The only operations ExplicitPtr supports are:
+  //
+  //   - Dereferencing with operator*.
+  //   - Indirection with operator->.
+  //   - Contextual conversion to bool with operator bool.
+  //
+  // If you require a raw pointer from an ExplicitPtr, you must use explicit syntax such as `&*p` or
+  // `p.operator->()`, or `static_cast<T*>(p)`.
+  //
+  // The purpose of this class is to provide greater compile-time safety for the variable declared
+  // by the `KJ_IF_MAYBE(p, maybe)` macro -- it is common for authors to forget that `p` is a
+  // pointer, and this confusion has been the frequent cause of subtle bugs with sometimes-severe
+  // consequences over the years. By making `p` an ExplicitPtr instead of a raw pointer,
+  // accidentally passing `p` to a function which accepts both a raw pointer and the pointed-to type
+  // (e.g., `kj::str()`, `kj::hashCode()`, and many more) becomes a compile-time error.
+  //
+  // This class is SFINAE-friendly, so that it may be used in function signatures wherever T* may be
+  // used.
+
+public:
+  ExplicitPtr(T* ptr): ref(ptr) {}
+  // A raw pointer can be implicitly converted to ExplicitPtr.
+
+  T& operator*() { return *ref; }
+  T* operator->() { return ref; }
+  const T& operator*() const { return *ref; }
+  const T* operator->() const { return ref; }
+  // ExplicitPtr supports pointer-like dereference/indirection operations.
+
+  explicit operator T*() { return ref; }
+  explicit operator const T*() const { return ref; }
+  // An ExplicitPtr may be explicitly converted back to its raw pointer.
```

Comment:
Is this actually useful, vs. `*&`?

Reply:
It might not be -- I think I put it in for symmetry with the other two types that `readMaybe()` returns. But I suppose those other two types have other reasons for being convertible. I'll remove.